contrib/libs/silk/contrib/librseq/doc/man/rseq.2
.\" SPDX-FileCopyrightText: 2015-2023 Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
.\"
.\" SPDX-License-Identifier: Linux-man-pages-copyleft
.\"
.TH rseq 2 (date) "Linux man-pages (unreleased)"
.SH NAME
rseq \- restartable sequences system call
.SH LIBRARY
Standard C library
.RI ( libc ", " \-lc )
.SH SYNOPSIS
.nf
.PP
.BR "#include <linux/rseq.h>" "       /* Definition of " RSEQ_* " constants */"
.BR "#include <sys/syscall.h>" "      /* Definition of " SYS_* " constants */"
.B #include <unistd.h>
.PP
.BI "int syscall(SYS_rseq, struct rseq *" rseq ", uint32_t " rseq_len ,
.BI "            int " flags ", uint32_t " sig );
.fi
.PP
.IR Note :
glibc provides no wrapper for
.BR rseq (),
necessitating the use of
.BR syscall (2).
.SH DESCRIPTION
The
.BR rseq ()
ABI accelerates specific user-space operations by registering a
per-thread data structure shared between kernel and user-space.
This data structure can be read from or written to by user-space to skip
otherwise expensive system calls.
.PP
A restartable sequence is a sequence of instructions
guaranteed to be executed atomically with respect to
other threads and signal handlers on the current CPU.
If its execution does not complete atomically,
the kernel changes the execution flow by jumping to an abort handler
defined by user-space for that restartable sequence.
.PP
Using restartable sequences requires to register a
.BR rseq ()
ABI per-thread data structure
.RB ( "struct rseq" )
through the
.BR rseq ()
system call.
Only one
.BR rseq ()
ABI can be registered per thread, so user-space libraries and
applications must follow a user-space ABI defining how to share this
resource.
The ABI defining how to share this resource between applications and
libraries is defined by the C library.
Allocation of the per-thread
.BR rseq ()
ABI and its registration to the kernel is handled by glibc since version
2.35.
.PP
The
.BR rseq ()
ABI per-thread data structure contains a
.I rseq_cs
field which points to the currently executing critical section.
For each thread, a single rseq critical section can run at any given
point.
Each critical section need to be implemented in assembly.
.PP
The
.BR rseq ()
ABI accelerates user-space operations on per-cpu data by defining a
shared data structure ABI between each user-space thread and the kernel.
.PP
It allows user-space to perform update operations on per-cpu data
without requiring heavy-weight atomic operations.
.PP
The term CPU used in this documentation refers to a hardware execution
context.
For instance, each CPU number returned by
.BR sched_getcpu ()
is a CPU.
The current CPU means the CPU on which the registered thread is running.
.PP
Restartable sequences are atomic with respect to preemption (making it
atomic with respect to other threads running on the same CPU),
as well as signal delivery (user-space execution contexts nested over
the same thread).
They either complete atomically with respect to preemption on the
current CPU and signal delivery, or they are aborted.
.PP
Restartable sequences are suited for update operations on per-cpu data.
.PP
Restartable sequences can be used on data structures shared between threads
within a process,
and on data structures shared between threads across different
processes.
.PP
Some examples of operations that can be accelerated or improved by this ABI:
.IP \(bu 3
Memory allocator per-cpu free-lists,
.IP \(bu 3
Querying the current CPU number,
.IP \(bu 3
Incrementing per-CPU counters,
.IP \(bu 3
Modifying data protected by per-CPU spinlocks,
.IP \(bu 3
Inserting/removing elements in per-CPU linked-lists,
.IP \(bu 3
Writing/reading per-CPU ring buffers content.
.IP \(bu 3
Accurately reading performance monitoring unit counters with respect to
thread migration.
.PP
Restartable sequences must not perform system calls.
Doing so may result in termination of the process by a segmentation
fault.
.PP
The
.I rseq
argument is a pointer to the thread-local
.B struct rseq
to be shared between kernel and user-space.
.PP
The structure
.B struct rseq
is an extensible structure.
Additional feature fields can be added in future kernel versions.
Its layout is as follows:
.TP
.B Structure alignment
This structure is aligned on either 32-byte boundary,
or on the alignment value returned by
.IR getauxval ()
invoked with
.B AT_RSEQ_ALIGN
if the structure size differs from 32 bytes.
.TP
.B Structure size
This structure size needs to be at least 32 bytes.
It can be either 32 bytes,
or it needs to be large enough to hold the result of
.IR getauxval ()
invoked with
.BR AT_RSEQ_FEATURE_SIZE .
Its size is passed as parameter to the
.BR rseq ()
system call.
.in +4n
.IP
.EX
#include <linux/rseq.h>

struct rseq {
    __u32 cpu_id_start;
    __u32 cpu_id;
    union {
        /* ... */
    } rseq_cs;
    __u32 flags;
    __u32 node_id;
    __u32 mm_cid;
} __attribute__((aligned(32)));
.EE
.in
.TP
.B Fields
.RS
.TP
.I cpu_id_start
Always-updated value of the CPU number on which the registered thread is
running.
Its value is guaranteed to always be a possible CPU number,
even when
.BR rseq ()
is not registered.
Its value should always be confirmed by reading the cpu_id field before
user-space performs any side-effect
(e.g. storing to memory).
.IP
This field is always guaranteed to hold a valid CPU number in the range
[ 0 ..  nr_possible_cpus - 1 ].
It can therefore be loaded by user-space
and used as an offset in per-cpu data structures
without having to check whether its value is within the valid bounds
compared to the number of possible CPUs in the system.
.IP
Initialized by user-space to a possible CPU number (e.g., 0),
updated by the kernel for threads registered with
.BR rseq ().
.IP
For user-space applications executed on a kernel without
.BR rseq ()
support,
the cpu_id_start field stays initialized at 0,
which is indeed a valid CPU number.
It is therefore valid to use it as an offset in per-cpu data structures,
and only validate whether it's actually the current CPU number by
comparing it with the cpu_id field within the rseq critical section.
If the kernel does not provide
.BR rseq ()
support, that cpu_id field stays initialized at -1,
so the comparison always fails, as intended.
.IP
This field should only be read by the thread which registered this data
structure.
Aligned on 32-bit.
.IP
It is up to user space to implement a fall-back mechanism for scenarios where
.BR rseq ()
is not available.
.TP
.I cpu_id
Always-updated value of the CPU number on which the registered thread is
running.
Initialized by user-space to -1,
updated by the kernel for threads registered with
.BR rseq ().
.IP
This field should only be read by the thread which registered this data
structure.
Aligned on 32-bit.
.TP
.I rseq_cs
The rseq_cs field is a pointer to a
.BR "struct rseq_cs" .
Is is NULL when no rseq assembly block critical section is active for
the registered thread.
Setting it to point to a critical section descriptor
.RB ( "struct rseq_cs")
marks the beginning of the critical section.
.IP
Initialized by user-space to NULL.
.IP
Updated by user-space, which sets the address of the currently
active rseq_cs at the beginning of assembly instruction sequence
block,
and set to NULL by the kernel when it restarts an assembly instruction
sequence block,
as well as when the kernel detects that it is preempting or delivering a
signal outside of the range targeted by the rseq_cs.
Also needs to be set to NULL by user-space before reclaiming memory that
contains the targeted
.BR "struct rseq_cs" .
.IP
Read and set by the kernel.
.IP
This field should only be updated by the thread which registered this
data structure.
Aligned on 64-bit.
.TP
.I flags
Flags indicating the restart behavior for the registered thread.
This is mainly used for debugging purposes.
Can be a combination of:
.RS
.TP
.B RSEQ_CS_FLAG_NO_RESTART_ON_PREEMPT
Inhibit instruction sequence block restart on preemption for this
thread.
This flag is deprecated since Linux 6.1.
.TP
.B RSEQ_CS_FLAG_NO_RESTART_ON_SIGNAL
Inhibit instruction sequence block restart on signal delivery for this
thread.
This flag is deprecated since Linux 6.1.
.TP
.B RSEQ_CS_FLAG_NO_RESTART_ON_MIGRATE
Inhibit instruction sequence block restart on migration for this thread.
This flag is deprecated since Linux 6.1.
.RE
.IP
Initialized by user-space, used by the kernel.
.TP
.I node_id
Always-updated value of the current NUMA node ID.
.IP
Initialized by user-space to 0.
.IP
Updated by the kernel.
Read by user-space with single-copy atomicity semantics.
This field should only be read by the thread which registered
this data structure.
Aligned on 32-bit.
.TP
.I mm_cid
Contains the current thread's concurrency ID
(allocated uniquely within a memory map).
.IP
Updated by the kernel.
Read by user-space with single-copy atomicity semantics.
This field should only be read by the thread which registered this data
structure.
Aligned on 32-bit.
.IP
This concurrency ID is within the possible cpus range,
and is temporarily (and uniquely) assigned while threads are actively
running within a memory map.
If a memory map has fewer threads than cores,
or is limited to run on few cores concurrently through sched affinity or
cgroup cpusets,
the concurrency IDs will be values close to 0,
thus allowing efficient use of user-space memory for per-cpu data
structures.
.RE
.PP
The layout of
.B struct rseq_cs
version 0 is as follows:
.TP
.B Structure alignment
This structure is aligned on 32-byte boundary.
.TP
.B Structure size
This structure has a fixed size of 32 bytes.
.in +4n
.IP
.EX
#include <linux/rseq.h>

struct rseq_cs {
    __u32   version;
    __u32   flags;
    __u64   start_ip;
    __u64   post_commit_offset;
    __u64   abort_ip;
} __attribute__((aligned(32)));
.EE
.in
.TP
.B Fields
.RS
.TP
.I version
Version of this structure.
Should be initialized to 0.
.TP
.I flags
.RS
Flags indicating the restart behavior of this structure.
Can be a combination of:
.TP
.B RSEQ_CS_FLAG_NO_RESTART_ON_PREEMPT
Inhibit instruction sequence block restart on preemption for this
critical section.
This flag is deprecated since Linux 6.1.
.TP
.B RSEQ_CS_FLAG_NO_RESTART_ON_SIGNAL
Inhibit instruction sequence block restart on signal delivery for this
critical section.
This flag is deprecated since Linux 6.1.
.TP
.B RSEQ_CS_FLAG_NO_RESTART_ON_MIGRATE
Inhibit instruction sequence block restart on migration for this
critical section.
This flag is deprecated since Linux 6.1.
.RE
.TP
.I start_ip
Instruction pointer address of the first instruction of the sequence of
consecutive assembly instructions.
.TP
.I post_commit_offset
Offset (from start_ip address) of the address after the last instruction
of the sequence of consecutive assembly instructions.
.TP
.I abort_ip
Instruction pointer address where to move the execution flow in case of
abort of the sequence of consecutive assembly instructions.
.RE
.PP
The
.I rseq_len
argument is the size of the
.B struct rseq
to register.
.PP
The
.I flags
argument is 0 for registration, and
.B RSEQ_FLAG_UNREGISTER
for unregistration.
.PP
The
.I sig
argument is the 32-bit signature to be expected before the abort
handler code.
.PP
A single library per process should keep the
.B struct rseq
in a per-thread data structure.
The
.I cpu_id
field should be initialized to -1, and the
.I cpu_id_start
field should be initialized to a possible CPU value (typically 0).
.PP
Each thread is responsible for registering and unregistering its
.BR "struct rseq" .
No more than one
.B struct rseq
address can be registered per thread at a given time.
.PP
Reclaim of
.B struct rseq
object's memory must only be done after either an explicit rseq
unregistration is performed or after the thread exits.
.PP
In a typical usage scenario, the thread registering the
.B struct rseq
will be performing loads and stores from/to that structure.
It is however also allowed to read that structure from other threads.
The
.B struct rseq
field updates performed by the kernel provide relaxed atomicity
semantics (atomic store, without memory ordering),
which guarantee that other threads performing relaxed atomic reads
(atomic load, without memory ordering) of the cpu number fields will
always observe a consistent value.
.SH RETURN VALUE
A return value of 0 indicates success.
On error, \-1 is returned, and
.I errno
is set appropriately.
.SH ERRORS
.TP
.B EINVAL
Either
.I flags
contains an invalid value, or
.I rseq
contains an address which is not appropriately aligned, or
.I rseq_len
contains an incorrect size, or restartable sequence is already
registered for this thread and at least one argument differs from the
active registration.
.TP
.B ENOSYS
The
.BR rseq ()
system call is not implemented by this kernel.
.TP
.B EFAULT
.I rseq
is an invalid address.
.TP
.B EBUSY
Restartable sequence is already registered for this thread and all arguments
are the same as the active registration.
.TP
.B EPERM
The
.I sig
argument on unregistration does not match the signature received
on registration.
.SH VERSIONS
The
.BR rseq ()
system call was added in Linux 4.18.
.SH STANDARDS
.BR rseq ()
is Linux-specific.
.SH SEE ALSO
.BR sched_getcpu (3) ,
.BR membarrier (2) ,
.BR getauxval (3)


contrib/libs/silk/contrib/librseq/src/rseq.c
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2016 Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
// SPDX-FileCopyrightText: 2026 Michael Jeanson <mjeanson@efficios.com>

#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif
#include <errno.h>
#include <sched.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <syscall.h>
#include <assert.h>
#include <signal.h>
#include <limits.h>
#include <dlfcn.h>
#include <stddef.h>
#include <stdint.h>
#include <sys/auxv.h>
#include <linux/auxvec.h>

#include <rseq/rseq.h>
#include "smp.h"

#ifndef AT_RSEQ_FEATURE_SIZE
# define AT_RSEQ_FEATURE_SIZE		27
#endif

#ifndef AT_RSEQ_ALIGN
# define AT_RSEQ_ALIGN			28
#endif


/*
 * Private internal variables.
 */

static pthread_mutex_t init_lock = PTHREAD_MUTEX_INITIALIZER;
static int init_done;

static const ptrdiff_t *libc_rseq_offset_p;
static const unsigned int *libc_rseq_size_p;
static const unsigned int *libc_rseq_flags_p;


/*
 * Public API variables.
 */

/* Offset from the thread pointer to the rseq area. */
ptrdiff_t rseq_offset = PTRDIFF_MIN;

/* Size of the active rseq features. 0 if the registration failed. */
unsigned int rseq_size = -1U;

/* Flags used at rseq registration. */
unsigned int rseq_flags = 0;


/*
 * Private util functions.
 */

/* rseq syscall wrapper. */
static
int sys_rseq(struct rseq_abi *rseq_abi, uint32_t rseq_len,
		    int flags, uint32_t sig)
{
	return syscall(__NR_rseq, rseq_abi, rseq_len, flags, sig);
}

/* getcpu syscall wrapper. */
static
int sys_getcpu(unsigned int *cpu, unsigned int *node)
{
	return syscall(__NR_getcpu, cpu, node, NULL);
}

/*
 * Return the feature size supported by the kernel.
 *
 * Depending on the value returned by getauxval(AT_RSEQ_FEATURE_SIZE):
 *
 * 0:   Return RSEQ_ABI_ORIG_FEATURE_SIZE (20)
 * > 0: Return the value from getauxval(AT_RSEQ_FEATURE_SIZE).
 *
 * It never returns a value below RSEQ_ABI_ORIG_FEATURE_SIZE.
 */
static
unsigned int get_rseq_kernel_feature_size(void)
{
	unsigned long auxv_rseq_feature_size = getauxval(AT_RSEQ_FEATURE_SIZE);

	if (auxv_rseq_feature_size)
		return auxv_rseq_feature_size;
	else
		return RSEQ_ABI_ORIG_FEATURE_SIZE;
}


/*
 * Public API functions.
 */

/*
 * Initialize librseq, must be called once per process.
 */
int rseq_init(void)
{
	/*
	 * Ensure initialization is only done once. Use load-acquire to
	 * observe the initialization performed by a concurrently
	 * running thread.
	 */
	if (rseq_smp_load_acquire(&init_done))
		return RSEQ_INIT_OK;

	/*
	 * Take the mutex, check the initialization flag again and atomically
	 * set it to ensure we are the only thread doing the initialization.
	 */
	pthread_mutex_lock(&init_lock);
	if (rseq_smp_load_acquire(&init_done))
		goto unlock_ok;

	/*
	 * Get the libc rseq public symbols, all 3 are required for a
	 * successful initialization.
	 */
	libc_rseq_offset_p = dlsym(RTLD_NEXT, "__rseq_offset");
	libc_rseq_size_p = dlsym(RTLD_NEXT, "__rseq_size");
	libc_rseq_flags_p = dlsym(RTLD_NEXT, "__rseq_flags");
	if (!libc_rseq_size_p || !libc_rseq_offset_p || !libc_rseq_flags_p) {
		pthread_mutex_unlock(&init_lock);
		return RSEQ_INIT_ERROR_MISSING_SYMBOLS;
	}

	/*
	 * Older versions of Glibc expose a rseq feature size of 32 bytes even
	 * though the kernel only supported 20 bytes initially. Glibc 2.40
	 * exposes a fixed feature size of 20 bytes, while still allocating a
	 * 32 bytes area.
	 *
	 * Treat both 32 and 20 bytes as special-cases using the following
	 * value as active feature size:
	 *
	 *   rseq_size = min(32, max(20, getauxval(AT_RSEQ_FEATURE_SIZE)));
	 *
	 * Otherwise, use the rseq_size from libc directly.
	 */
	switch (*libc_rseq_size_p) {
	case RSEQ_ABI_ORIG_FEATURE_SIZE:	/* Fallthrough. */
	case RSEQ_ABI_ORIG_ALLOC_SIZE:
	{
		unsigned int rseq_kernel_feature_size = get_rseq_kernel_feature_size();

		if (rseq_kernel_feature_size < RSEQ_ABI_ORIG_ALLOC_SIZE)
			rseq_size = rseq_kernel_feature_size;
		else
			rseq_size = RSEQ_ABI_ORIG_ALLOC_SIZE;
		break;
	}
	default:
		/* Otherwise just use the __rseq_size from libc as rseq_size. */
		rseq_size = *libc_rseq_size_p;
		break;
	}

	/* Copy the libc rseq offset and flags values. */
	rseq_offset = *libc_rseq_offset_p;
	rseq_flags = *libc_rseq_flags_p;

	/*
	 * Set init_done with store-release, to make sure concurrently
	 * running threads observe the initialized state.
	 */
	rseq_smp_store_release(&init_done, 1);
unlock_ok:
	pthread_mutex_unlock(&init_lock);
	return RSEQ_INIT_OK;
}

/*
 * Returns true if rseq is supported. The query types are:
 *
 *   RSEQ_AVAILABLE_QUERY_KERNEL:
 *     Returns true if the rseq syscall is available.
 *
 *   RSEQ_AVAILABLE_QUERY_LIBC:
 *     Returns true if the libc exposes the rseq symbols.
 */
bool rseq_available(unsigned int query)
{
	int rc;

	switch (query) {
	case RSEQ_AVAILABLE_QUERY_KERNEL:
		rc = sys_rseq(NULL, 0, 0, 0);
		if (rc != -1)
			abort();
		switch (errno) {
		case ENOSYS:
			break;
		case EINVAL:
			return true;
		default:
			abort();
		}
		break;
	case RSEQ_AVAILABLE_QUERY_LIBC:
		libc_rseq_offset_p = dlsym(RTLD_NEXT, "__rseq_offset");
		libc_rseq_size_p = dlsym(RTLD_NEXT, "__rseq_size");
		libc_rseq_flags_p = dlsym(RTLD_NEXT, "__rseq_flags");
		if (libc_rseq_offset_p && libc_rseq_size_p && libc_rseq_flags_p)
			return true;
		break;
	default:
		break;
	}
	return false;
}

/*
 * Slow fallback to get the current CPU number.
 */
int32_t rseq_fallback_current_cpu(void)
{
	int32_t cpu;

	cpu = sched_getcpu();
	if (cpu < 0) {
		perror("sched_getcpu()");
		abort();
	}
	return cpu;
}

/*
 * Slow fallback to get the current node number.
 */
int32_t rseq_fallback_current_node(void)
{
	uint32_t cpu_id, node_id;
	int ret;

	ret = sys_getcpu(&cpu_id, &node_id);
	if (ret) {
		perror("sys_getcpu()");
		return ret;
	}
	return (int32_t) node_id;
}

/*
 * Returns the max_nr_cpus auto-detected at pool creation when invoked
 * with @nr_max_cpus=0 argument.
 */
int rseq_get_max_nr_cpus(void)
{
	return get_possible_cpus_array_len();
}
